lib/collectors/global_prometheus_collector.rb

```diff
@@ -0,0 +1,40 @@
+require "prometheus_exporter"
+require "prometheus_exporter/server"
+
+# Load Rails models as not automatically included for Prometheus Exporter
+require File.expand_path("../../config/environment", __dir__) unless defined? Rails
+
+module Collectors
+  class GlobalPrometheusCollector < PrometheusExporter::Server::TypeCollector
+    def type
+      "app_global"
+    end
+
+    def metrics
+      token_expiry_timestamp = PrometheusExporter::Metric::Gauge.new("signon_api_user_token_expiry_timestamp_seconds", "Timestamp when API User token expires")
+
+      # Cache metric to prevent needless expensive calls to the database
+      token_expiry_info = Rails.cache.fetch("token_expiry_info", expires_in: 1.hour) do
```

Comment:
How often will this code be called? Do we know whether a 1 hour cache is sufficient to prevent those calls? Also - have you had a look to see how expensive these calls are in practice, I wonder if we could avoid the complication of caching if they're not expensive to run.

Reply:
So Prometheus will scrape this endpoint every minute. Individual token expiry timestamps never change. Changes in the data would be around tokens being created or deleted - which happen very seldom (a generous estimate would be maybe a couple times a month). Without any caching we're going to be doing a great deal of unnecessary expensive database calls.

1 hour was trade of minimising expensive calls, vs the freshness of the data we can see in our monitoring system - but probably could even be okay bumping caching up to a day.

Reply:
Is the once-per-minute frequency of calls configurable? Would scraping less often would be an alternative to caching (given that there isn't anything to expire that cache other than the expiry time).

Reply:
Not really - the /metrics endpoint has other metrics like http requests total, which we do want per minute resolution on